### DISCUSSION TITLE
[Bug]: header and footer image not loading

### DISCUSSION THREAD

**Discussion by RVP04** [spatie/laravel-pdf#71] [Q&A]
### What happened?

header blade
```
<!DOCTYPE html>
<html lang="en">
<head>
    <style>
        .header {
            font-family: "Courier New";
            position: fixed;
            top: 10px;
            left: 0;
            right: 20px;
            text-align: center;
        }
    </style>
</head>
<body>
<div class="header">
    <img src="{{ asset('img/gray-logo.png') }}" width="160" height="70" style="float: right" alt="">
    <p style="display: flex;margin-top: 40px; font-size: 12pt;font-weight: bold;align-items: center;justify-content: center">{{ $reportTitle ?? '' }}</p>
</div>
</body>
</html>
```

footer.blade.php
```
<!DOCTYPE html>
<html>

<head>
    <style>
        .footer {
            position: absolute;
            left: 0px;
            right: 0px;
            height: 10px;
            float: left;
            font-size: 7pt;
            border-top: 1px solid black;
            padding-top: 10px;
            font-family: 'Courier New';
        }

        .page-number {
            overflow: hidden;
            float: right;
            margin-right: 40px;
        }

        .report {
            overflow: hidden;
            width: 250px;
            float: left;
            margin-left: 10px;
        }
    </style>
</head>

<body>
<div class="footer">
    <div class="report">Page @pageNumber of @totalPages</div>
    <div class="page-number"><img src="{{public_path('img/broad-logo.png')}}" width="100" height="20"></div>
</div>
</body>
</html>
```

### How to reproduce the bug

PDF budlider
```
return Pdf::view('reports.pdf.status_subject_wise', [
            'subjects' => $subjects
        ])->headerView('layout.reports.header', ['reportTitle' => 'Feed Back Status - Subject Wise'])
            ->footerView('layout.reports.footer')
            ->format('a4')
            ->orientation(Orientation::Landscape)
            ->margins(2.5, 0.5, 1.2, 0.5, 'cm')
            ->name('Feedback Status - Subject Wise.pdf');
```
1. Image is not loading using ``` asset() helper``` or ``` public_path()``` there is no sign of image in the document. When use ``` alt ``` attribute it is not showing image not loaded
2. when use ``` asset()``` first page logo not showing (image not loaded it is not coming see the image) and if public path is use just a box appears

```asset```
<img width="307" alt="Screenshot 2024-02-03 at 4 34 56 AM" src="https://github.com/spatie/laravel-pdf/assets/8185511/7b8ff2d1-a9af-4345-96b1-c926221edc4a">
```public_path```
<img width="191" alt="Screenshot 2024-02-03 at 4 35 21 AM" src="https://github.com/spatie/laravel-pdf/assets/8185511/914251be-265e-477a-a51f-ab2186ef69e3">

need a solution for this isseues

### Package Version

1.1.2

### PHP Version

8.3.0

### Laravel Version

10.40

### Which operating systems does with happen with?

macOS

### Notes
No use of #44 discussion 

_No response_

**Comment by RVP04** [comment]
Need help on this please @freekmurze @crynobone @ArielMejiaDev

**Comment by ArielMejiaDev** [comment]
Hi @RVP04 the images could fail to load for a variety of reasons here my suggestion:

Instead of setting a public_path like this 

```php
<div class="page-number"><img src="{{public_path('img/broad-logo.png')}}" width="100" height="20"></div>
```

an absolute path to an external service eg: (s3) could make it work

```php
// config/brand.php

[
    'brand' => [
        'logo' => 'https://aws/my-bucket/images/brand/logo.png'
    ]
]

// then in your blade component

<div class="page-number"><img src="{{ config('brand.logo') }}" width="100" height="20"></div>
``` 

This should fix the issue in all environments, remember that locally a relative path should not work and as you could have more than one environment this could lead to a lot of work just to show an image to check environments, etc.... it is easier to direclty load them externally.

**Reply by RVP04** [reply]
Instead of AWS can I use local path because I don't have plan to put this website in AWS @ArielMejiaDev

**Reply by ArielMejiaDev** [reply]
Take in mind that it was just an example, the key is to replace your **local path** and add instead a **absolute path**, this means your image `src` should become from an external URL, you can even host your image in github as an example, but it could be with any other service, eg:

- [Host your image in github](https://medium.com/@jeftachibiya360/how-to-host-images-for-your-website-on-github-a98d917284c5)

**Reply by RVP04** [reply]
Okay sir let me try and update you

**Reply by RVP04** [reply]
@ArielMejiaDev i tried all possibilities it is not working sir, @freekmurze please you try and see it will helpful for all

**Reply by RVP04** [reply]
when i use normal view image is loading, only in pdf image is not loading
```
<img src="{{ asset('img/gray-logo.png') }}" width="160" height="70" style="float: right" alt="">
```

**Reply by RVP04** [reply]
i also tried this one sir.
```
<!DOCTYPE html>
<html lang="en">
<head>
    <script src="https://cdn.tailwindcss.com"></script>
    <style>
        .header {
            font-family: "Courier New";
            position: fixed;
            top: 10px;
            left: 0;
            right: 20px;
            text-align: center;
        }

        img {
            background-image: {{ url('img/gray-logo.png') }};
        }
    </style>
</head>
<body>
<div class="header">
    <img width="155" height="70" style="float: right">
    <p style="display: flex;margin-top: 40px; font-size: 12pt;font-weight: bold;align-items: center;justify-content: center">{{ $reportTitle ?? '' }}</p>
</div>
</body>
</html>
```
i tested on the content it is working. Only on the headerView() and footerView() image is not showing

**Comment by ArielMejiaDev** [comment]
@RVP04 @freekmurze I upload a demo with header and footer images working for relative and absolute paths

[Here is the demo project link](https://github.com/ArielMejiaDev/laravel-pdf-demo)

## Video

https://www.loom.com/share/8912fba845214a1d93aefc99fcf93b93?sid=46351d44-b10f-4830-8841-8cbbd61e1b39

I also make a PR GH-79 to solve this with a package solution, please let me know your feedback

**Reply by RVP04** [reply]
@ArielMejiaDev appreciated. I am waiting for @freekmurze to check and push the pr to branch

**Comment by jonaaix** [comment]
I have the same issue, could solve it by converting the image URL to base64 first.
In the header/footer, any urls are not working. Thus also not tailwind.

```blade
<header style="width: 100%; display: flex; justify-content: end; padding-right: 1cm;">
   @php
      $logoUrl = $tenant->logo?->getLargeUrl();
      $logo64 = \App\Services\HelperService::imageUrlToBase64($logoUrl);
   @endphp

   <img style="width: 300px;" src="{{ $logo64 }}" alt="Logo">
</header>
```

Helper function:
```php
public static function imageUrlToBase64(string $imageUrl): string {
  $imageContents = file_get_contents($imageUrl);
  if ($imageContents === false) {
     throw new \RuntimeException("Failed to fetch the image from URL: {$imageUrl}");
  }

  $mimeType = finfo_buffer(finfo_open(FILEINFO_MIME_TYPE), $imageContents);
  $base64 = base64_encode($imageContents);

  return "data:{$mimeType};base64,{$base64}";
}
```
